engines/director/score.cpp

```diff
@@ -1344,123 +1357,273 @@ void Score::playQueuedSound() {
+// data after the jump!
+void Score::rebuildChannelData(int frameNum) {
+	// Builds channel data from frame 1 to frame.
+	if (frameNum > (int)_numFrames) {
+		warning("Score::rebuildChannelData(): frameNum %d is greater than total frames %d", frameNum, _numFrames);
+		return;
+	}
 
+	memset(_channelData, 0, kChannelDataSize);
+
+	// Lock variables
+	int curFrameNumber = _curFrameNumber;
+	Frame *frame = _currentFrame;
+	_currentFrame = nullptr; // To avoid later deletion of frame inside renderOneFrame()
+
+	_framesStream->seek(_frameOffsets[1]); // Seek to frame 1
+	for (int i = 1; i < frameNum; i++) {
+		readOneFrame(false);
+	}
+
+	// Unlock variables
+	delete _currentFrame; // Destroy the built frame
+	_curFrameNumber = curFrameNumber;
+	_currentFrame = frame;
+}
+
+bool Score::readOneFrame(bool saveOffset) {
+	uint16 channelSize;
+	uint16 channelOffset;
+
+	if (_framesStream->pos() < _framesStreamSize && !_framesStream->eos()) {
+		uint16 frameSize = _framesStream->readUint16();
+		assert(frameSize < _framesStreamSize);
+
+		debugC(3, kDebugLoading, "++++++++++ score load frame (frameSize %d) saveOffset %d", frameSize, saveOffset);
+		if (debugChannelSet(8, kDebugLoading)) {
+			_framesStream->hexdump(frameSize);
+		}
 		if (frameSize > 0) {
 			Frame *frame = new Frame(this, _numChannelsDisplayed);
-			size -= frameSize;
 			frameSize -= 2;
 
 			while (frameSize != 0) {
 
 				if (_vm->getVersion() < 400) {
-					channelSize = stream.readByte() * 2;
-					channelOffset = stream.readByte() * 2;
+					channelSize = _framesStream->readByte() * 2;
+					channelOffset = _framesStream->readByte() * 2;
 					frameSize -= channelSize + 2;
 				} else {
-					channelSize = stream.readUint16();
-					channelOffset = stream.readUint16();
+					channelSize = _framesStream->readUint16();
+					channelOffset = _framesStream->readUint16();
 					frameSize -= channelSize + 4;
 				}
 
 				assert(channelOffset + channelSize < kChannelDataSize);
-				stream.read(&channelData[channelOffset], channelSize);
+				_framesStream->read(&_channelData[channelOffset], channelSize);
 			}
 
-			Common::MemoryReadStreamEndian *str = new Common::MemoryReadStreamEndian(channelData, ARRAYSIZE(channelData), stream.isBE());
+			Common::MemoryReadStreamEndian *str = new Common::MemoryReadStreamEndian(_channelData, ARRAYSIZE(_channelData), _framesStream->isBE());
 			// str->hexdump(str->size(), 32);
-			frame->readChannels(str, version);
+			frame->readChannels(str, _version);
 			delete str;
 			// Precache the current FPS tempo, as this carries forward to frames to the right
 			// of the instruction.
 			// Delay type tempos (e.g. wait commands, delays) apply to only a single frame, and are ignored here.
 			if (frame->_tempo && frame->_tempo <= 120)
-				currentTempo = frame->_tempo;
-			frame->_scoreCachedTempo = frame->_tempo ? frame->_tempo : currentTempo;
+				_currentTempo = frame->_tempo;
```

Comment:
This will not work if we seek back, right? This is another thing to test with the original, e.g. imagine the following movie:

```
frame 1: tempo 30
...
frame 10: tempo 40
frame 15: go to frame 5
```
The question is, will it keep tempo 40 after jumping or reset to 30? It is easy to test, just print out Tempo and frame number at every frame and see what sticks.